apollo-router/src/services/layers/persisted_queries/manifest_poller.rs

```diff
@@ -18,102 +22,303 @@ use crate::uplink::persisted_queries_manifest_stream::PersistedQueriesManifestCh
+    },
+    LogUnlessInSafelist {
+        safelist: FreeformGraphQLSafelist,
+        apq_enabled: bool,
+    },
+}
+
+/// Describes what the router should do for a given request: allow it, deny it
+/// with an error, or allow it but log the operation as unknown.
+pub(crate) enum FreeformGraphQLAction {
+    Allow,
+    Deny,
+    AllowAndLog,
+}
+
+impl FreeformGraphQLBehavior {
+    fn action_for_freeform_graphql(
+        &self,
+        body_from_request: &str,
+        ast: SyntaxTree,
+    ) -> FreeformGraphQLAction {
+        match self {
+            FreeformGraphQLBehavior::AllowAll { .. } => FreeformGraphQLAction::Allow,
+            FreeformGraphQLBehavior::DenyAll => FreeformGraphQLAction::Deny,
+            FreeformGraphQLBehavior::AllowIfInSafelist { safelist, .. } => {
+                if safelist.is_allowed(body_from_request, ast) {
+                    FreeformGraphQLAction::Allow
+                } else {
+                    FreeformGraphQLAction::Deny
+                }
+            }
+            FreeformGraphQLBehavior::LogUnlessInSafelist { safelist, .. } => {
+                if safelist.is_allowed(body_from_request, ast) {
+                    FreeformGraphQLAction::Allow
+                } else {
+                    FreeformGraphQLAction::AllowAndLog
+                }
+            }
+        }
+    }
+}
+
+/// The normalized bodies of all operations in the PQ manifest.
+///
+/// Normalization currently consists of:
+/// - Sorting the top-level definitions (operation and fragment definitions)
+///   deterministically.
+/// - Printing the AST using apollo-encoder's default formatting (ie,
+///   normalizing all ignored characters such as whitespace and comments).
+///
+/// Sorting top-level definitions is important because common clients such as
+/// Apollo Client Web have modes of use where it is easy to find all the
+/// operation and fragment definitions at build time, but challenging to
+/// determine what order the client will put them in at run time.
+///
+/// Normalizing ignored characters is helpful because being strict on whitespace
+/// is more likely to get in your way than to aid in security --- but more
+/// importantly, once we're doing any normalization at all, it's much easier to
+/// normalize to the default formatting instead of trying to preserve
+/// formatting.
+#[derive(Debug)]
+pub(crate) struct FreeformGraphQLSafelist {
+    normalized_bodies: HashSet<String>,
+}
+
+impl FreeformGraphQLSafelist {
+    fn new(manifest: &PersistedQueryManifest) -> Self {
+        let mut safelist = Self {
+            normalized_bodies: HashSet::new(),
+        };
+
+        for body in manifest.values() {
+            safelist.insert_from_manifest(body);
+        }
+
+        safelist
+    }
+
+    fn insert_from_manifest(&mut self, body_from_manifest: &str) {
+        self.normalized_bodies.insert(
+            self.normalize_body(body_from_manifest, Parser::new(body_from_manifest).parse()),
```

Comment:
is this a semi-permanent decision or one made in pursuit of rapid development? if the latter, let's make a comment to re-visit pulling a parser out of the compiler that's in context from the `query_analysis_layer`.

Reply:
@EverlastingBugstopper I'm not sure I follow. This is the function that runs when you load a manifest, not on an incoming request.

Reply:
oh shoot you're totally right, disregard